Subscription Block: Add success message

amustaque97 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.MD in a new commit as part of your PR. -->

Fixes #21187

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Updated `projects/plugins/jetpack/extensions/blocks/subscriptions` to support teaxtarea for `success_message` in the widget setttings.
* Updated `projects/plugins/jetpack/modules/subscriptions/views.php` to support `success_message` when passed as an attribute to the shortcode.
* Updated test cases of subscription form extension

#### Jetpack product discussion NA
<!-- If you're an Automattician, include a short link to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request to change what data or activity we track or use? NA
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that the context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Checkout to the PR branch
* Build subscription extension
* Enable jetpack subscription under discussion tab of jetpack settings
* Go to appearance -> customize -> widgets -> footer#1 -> add a legacy widget subscription block (if it not already transformed to subscription block and then transform it to subscription block)
* Select subscription block and click on show more settings and scroll down
* Edit success message and publish it.
* Open the site in incognito mode and then enter an email address
* Verify success message.

#### Screenrecording https://youtu.be/Z5cSgh5neQA

sdixon194 commented:
Thanks for the contribution @amustaque97 ! Is this PR still in progress or ready for review?

sdixon194 reviewed (changes requested):
This works for me! I have a few in line comments that I think we can tighten up. I'm also not an expert on block or react, so if someone else would like to take a look to make sure I'm not missing anything, that'd be appreciated too!

sdixon194 commented on projects/plugins/jetpack/extensions/blocks/subscriptions/controls.js:
```diff
@@ -245,6 +246,14 @@ export default function SubscriptionControls( {
+
+			<PanelBody title={ __( 'Success Message Text', 'jetpack' ) }>
+				<TextareaControl
+					value={ successMessage }
```
I think it's important for screen readers for the box to be labeled, but agreed that we don't want thins too cluttered/repetitive. What do you think about [setting hideLabelFromVision](https://developer.wordpress.org/block-editor/reference-guides/components/textarea-control/#hidelabelfromvision) to true, so it still shows up for screen readers, but won't appear visually?

Does that sound viable, @jeherve ?

jeherve commented:
Thank you. Could you run `git merge origin/master` in your branch one more time, after pulling the latest version of `master` in your fork?

amustaque97 commented:
> Thank you. Could you run `git merge origin/master` in your branch one more time, after pulling the latest version of `master` in your fork?

Take merge from the master branch. Test cases are passing.

jeherve reviewed (dismissed):
We're on a good path here. However, I just saw that you did not add a new deprecation so it can migrate to the latest block structure.

You can take a look at an example the last time we edited this block's structure: #18936 
And the official docs are available [here](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/).

Thanks!

amustaque97 commented:
Hi @jeherve 👋🏻, I had to revert few commits because of the mistakes I did here. I have addressed review comment. I have added a new deprecation version V6 and its test cases. Subscription test suite is completely passing. I believe this time it is correct. 

Can you please take a look at the PR again thoroughly? 

Thanks 😄

amustaque97 commented on projects/plugins/jetpack/extensions/blocks/subscriptions/controls.js:
```diff
@@ -245,6 +246,19 @@ export default function SubscriptionControls( {
 				/>
 			</PanelBody>
+
+			<PanelBody title={ __( 'Success Message Text', 'jetpack' ) }>
```
Yeah, I'm okay with `initialOpen` defaults to false. Make sense to me, checked it again. UI looks more consistent.

amustaque97 commented on projects/plugins/jetpack/extensions/blocks/subscriptions/deprecated/v6/save.js:
```diff
@@ -0,0 +1,149 @@
+		submit_button_classes: submitButtonClasses,
+		email_field_classes: emailFieldClasses,
+		show_only_email_and_button: true,
+		success_message: successMessage,
```
@samiff, I have addressed review comments. You can take a look at it. Can you please share what are the steps to test deprication. Till now I'm testing deprecation based on test cases and this [link](https://github.com/Automattic/jetpack/blob/d2e2f8e7dcc8bfa5489e9678705714f4ec2566c7/projects/plugins/jetpack/extensions/shared/test/block-fixtures.md)

samiff commented on projects/plugins/jetpack/extensions/blocks/subscriptions/deprecated/v6/index.js:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Deprecation reason
+ *
+ * Added a new input field successMessage.
```
```suggestion
 * Added new block attribute `successMessage`, which was already available to the shortcode.
```

samiff commented on projects/plugins/jetpack/extensions/blocks/subscriptions/deprecated/v6/index.js:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Internal dependencies
+ */
+import definedAttributes from '../v3/attributes';
+import save from './save';
```
Conceptually, the attributes and save method should be "identical" to what was present before the changes/deprecation occured. So I think here it'd be most correct to add a new `v6/attributes.js` file (that mirrors the attributes before adding the successMessage), and in the `v6/save.js` to import those attributes and mirror the `save.js` implementation before you started making changes - imports may need adjusting (e.g. the constants).

amustaque97 replied:
Hi, I have created a new file `attributes.js`. I don't need to change any constants. :)

samiff commented on projects/plugins/jetpack/extensions/blocks/subscriptions/deprecated/v6/index.js:
```diff
@@ -0,0 +1,25 @@
+ */
+export default {
+	attributes: definedAttributes,
+	migrate: attributes => {
```
I'm not thinking there needs to be a `migrate` method here. We aren't needing to modify previous attributes, we are simply adding a new attribute. Older block versions should still validate/save, and then they can optionally use the new `successMessage`. @jeherve Does that match your understanding when a new attribute is being added?

amustaque97 replied:
I got your idea, what are you're trying to say... Overall as far as i understand deprecation `v6` is nothing but current version of block master branch. Once my code gets merged it will be considered as new/updated version of the block. 

`migrate` method is only required when we are updating/changing existing block attribute. If we are adding a new block attribute we don't need `migrate` method.

samiff replied:
> migrate method is only required when we are updating/changing existing block attribute. If we are adding a new block attribute we don't need migrate method.

Right, unless for some reason we needed to alter past deprecated versions to include that new attribute. In this case `successMessage` has a default value and can optionally be customized. So the past versions don't need to include the new attribute in their migrate methods.

samiff reviewed (changes requested):
Thanks for adding those changes @amustaque97 I've left a couple comments - think we are almost there, just want to make sure on deprecation best practices.

samiff commented on projects/plugins/jetpack/extensions/blocks/subscriptions/deprecated/v6/save.js:
```diff
@@ -0,0 +1,147 @@
+export const DEFAULT_BORDER_WEIGHT_VALUE = 1;
+export const DEFAULT_PADDING_VALUE = 15;
+export const DEFAULT_SPACING_VALUE = 10;
+export const DEFAULT_FONTSIZE_VALUE = 16;
```
```suggestion
export const DEFAULT_FONTSIZE_VALUE = '16px';
```



I'm not sure where this version of `save.js` came from, but it appears to be missing a few of the changes you can see in the [copy on master currently](https://github.com/Automattic/jetpack/blob/d001afc08e7a70209cfdaafbbca14a91b0e65a25/projects/plugins/jetpack/extensions/blocks/subscriptions/save.js) (minus your `successMessage` changes of course). For example this `save.js` is missing the `buttonWidth` attribute.

amustaque97 replied:
I have made the required changes.

samiff commented on projects/plugins/jetpack/extensions/blocks/subscriptions/test/fixtures/jetpack__subscriptions.html:
```diff
@@ -1,3 +1,3 @@
-<!-- wp:jetpack/subscriptions {"showSubscribersTotal":true,"buttonOnNewLine":true,"fontSize":33,"customFontSize":33,"borderRadius":8,"borderWeight":3,"padding":18} -->
-<div class="wp-block-jetpack-subscriptions wp-block-jetpack-subscriptions__supports-newline wp-block-jetpack-subscriptions__use-newline wp-block-jetpack-subscriptions__show-subs">[jetpack_subscription_form show_subscribers_total="true" button_on_newline="true" custom_font_size="33" custom_border_radius="8" custom_border_weight="3" custom_padding="18" custom_spacing="10" submit_button_classes="has ...
+<!-- wp:jetpack/subscriptions {"subscribePlaceholder":"Enter your email address","showSubscribersTotal":true,"buttonOnNewLine":true,"submitButtonText":"Sign Up","fontSize":"33px","customFontSize":"33px","borderRadius":8,"borderWeight":3,"padding":18,"successMessage":"Volla, you subscribed to our blog!"} -->
```
Were these non-v6 files automatically generated for you? I tried generating the fixtures on my end but it only produced `deprecated-6` items. I'm thinking once the `save.js` is fixed up, you might try deleting the fixtures here and generating theme once more. Here is what my starting .html v6 looked like:

```html
<!-- wp:jetpack/subscriptions {"successMessage":"Custom success message text."} -->
<div class="wp-block-jetpack-subscriptions wp-block-jetpack-subscriptions__supports-newline">[jetpack_subscription_form show_subscribers_total="false" button_on_newline="false" custom_font_size="16px" custom_border_radius="0" custom_border_weight="1" custom_padding="15" custom_spacing="10" submit_button_classes="" email_field_classes="" show_only_email_and_button="true" success_message="Custom success message text."]</div>
<!-- /wp:jetpack/subscriptions -->
```

amustaque97 replied:
Non-v6 files were not generated in my case. It is only generating/modifying files for v6.

samiff reviewed (changes requested):
Looking good @amustaque97 ! just some cleanup in `v6/save.js` is needed and I would try redoing the fixtures once that is completed 👍  

On the sidebar of this PR I _think_ there should be an option to let Jetpack maintainers write to your branch if you don't mind so we don't keep having to ask for a master merge 😉

samiff reviewed (approved):
Those last changes look good, and the tests are passing from what I see 👍 

Master will need to be merged once more to bring this up to date.

amustaque97 commented:
Thank you @jeherve and @samiff for helping me with this. I can see this PR took long than expected, more than 2 months to be exact 😅 . Thanks once again for helping me out with this and patiently reviewing my every single commit and multiple code reviews. :heart: 

Please consider/ping me in other issues/PR related to deprecation implementation. Would love to get involved. 🚀

samiff commented:
Thanks for working on this @amustaque97 🎉   This change will ship out in Jetpack version 10.5 which is scheduled to release on: 2022-01-11


Internal: D71163-code marked for 10.5 release (sync with wpcom release)

jeherve commented:
Follow-up: #22193 seems to highlight an issue on WordPress.com Simple sites.
